components/raftstore/src/store/fsm/store.rs

```diff
@@ -2744,13 +2759,32 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> StoreFsmDelegate<'a, EK, ER
                 .swap(0, Ordering::Relaxed),
         );
 
-        stats.set_is_busy(
-            self.ctx
-                .global_stat
-                .stat
-                .is_busy
-                .swap(false, Ordering::Relaxed),
-        );
+        let store_is_busy = self
+            .ctx
+            .global_stat
+            .stat
+            .is_busy
+            .swap(false, Ordering::Relaxed);
+        let pending_recovery = {
+            (time::get_time().sec as u32).saturating_sub(start_time)
+                <= STORE_RECOVERY_DURATION.as_secs() as u32
+        };
+        let is_busy = if pending_recovery {
+            // If the store is busy in handling recovery when starting, it should not be
+            // treated as a normal store for balance. Only when the store is
+            // almost idle (no more pending regions on recovery), it can be
+            // regarded as candidates for balance.
+            let target_count = std::cmp::max(
+                1,
+                self.ctx.cfg.min_recovery_ready_region_percent * (stats.get_region_count() as u64)
+                    / 100,
+            );
+            let recovered_count = self.ctx.store_meta.lock().unwrap().recovered_peers_count;
```

Comment:
Maybe we can get `recovered_peers_count` at line 2723 (to save acquiring lock here)?